### ISSUE TITLE
Constructor parameters should match property names

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#33796]
In general, constructor arguments who initialize should match properties (casing aside). There are also cases where serializers will match properties to constructor parameters in order to construct immutable types. This would address that too.

**Category**: Design

**Comment by jeffhandley** [comment]
Estimates:
* Analyzer: Medium
* Fixer: Small

**Comment by jeffhandley** [comment]
We need to detect an inclusion approach; we want to focus on the JSON deserialization scenario.

**Comment by bartonjs** [comment]
This should only apply to constructors painted with the `[JsonConstructor]` serializer, but should be prepared to account for other enrollment scenarios.

* For each parameter name in the parameter list of an included constructor
  * If there is not a property on the type hierarchy at that level whose name is OrdinalIgnoreCase equal to the parameter name, report a diagnostic against the parameter (ideally just the name).

If the engine already has a "close match" name detector, then a fixer could suggest correcting a typo; but otherwise a fixer seems too open of a problem (unless other bright ideas surface).

**Comment by jeffhandley** [comment]
For existing violations in `runtime`, we likely would not apply the fix since changing parameter names would often be a breaking change.

**Comment by swtrse** [comment]
It should also considered that the JsonPropertyAttribute should be able to decorate constructor parameters.
This way the exact name in case sensitive format can be defined for the constructor parameters.

**Comment by danmosemsft** [comment]
@jeffhandley is this 5.0 work?

**Comment by jeffhandley** [comment]
@danmosemsft Moved to Future. Thanks.

We will target delivering this off-cycle soon after 5.0 since it will be good to get this out before folks start producing code that will violate the rule.

**Comment by psxvoid** [comment]
Hi,

May I "book" this issue as "up-for-grabs"?

**Comment by danmosemsft** [comment]
@psxvoid you are welcome, I assigned it to you. (@buyaa-n I am assuming we are not working on this since it's not assigned)

**Comment by buyaa-n** [comment]
> I am assuming we are not working on this since it's not assigned

That is right

**Comment by psxvoid** [comment]
Thank you.

I've spent some time on initial planning, and I'm wondering whether the following questions can be addressed before the implementation:

1. Should this analyzer also check field names? See note (3).
2. Should this analyzer be only applied to constructors marked as "JsonConstructor" from "System.Text.Json.Serialization" namespace, and not from "Newtonsoft.Json"?
3. What target namespace should be used for this analyzer? See note (5).

Notes:
(1) I'm assuming this analyzer is also must be applied to C# 9 Records (and structs). This is a valid usage:
> ```cs
> public record Person
> {
>    public string FirstName { get; init; }
>    public string LastName { get; init; }
>
>    [JsonConstructor]
>    public Person(string firstName, string lastName)
>      => (FirstName, LastName) = (firstName, lastName);
> }
> ```

(2) I'm assuming this analyzer should also be applied to VB.

(3) This [Migrate from Newtonsoft.Json to System.Text.Json - .NET | Microsoft Docs](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-5-0#public-and-non-public-fields) doc says:

> In `System.Text.Json`, use the `JsonSerializerOptions.IncludeFields` global setting or the `[JsonInclude]` attribute to include public fields when serializing or deserializing. For an example, see Include fields.

(4) It seems like uninitialized public properties are not a responsibility of this analyzer. Example:
> ```cs
> public class C
> {
>     public int P1 { get; }
>     public int? P2 { get; }
> 
>     [JsonConstructor]
>     public C(int p1)
>     {
>         this.P1 = p1;
>         // P2 is not initialized
>     }
> }
> ```

(5) It seems like this analyzer belongs to the "Design" category. As far as I can see, most of them reside in:
> `NetAnalyzers\Core\Microsoft.CodeQuality.Analyzers\ApiDesignGuidelines`

but this issue is also marked as `area-System.Runtime` label, and it could be also applicable to VB. Because of that, should it be placed here:

> `NetAnalyzers\Core\Microsoft.NetCore.Analyzers\Runtime`

**Comment by buyaa-n** [comment]
@psxvoid sorry for the late reply, saw your questions just today: 

> 1. Should this analyzer also check field names? See note (3).

 I don't think so, [JsonConstructor] attribute doesn't seem to apply/used for setting fields cc @bartonjs, @layomia   

> 2. Should this analyzer be only applied to constructors marked as "JsonConstructor" from "System.Text.Json.Serialization" namespace, and not from "Newtonsoft.Json"?

Yes, only from "System.Text.Json.Serialization" 

> 3. What target namespace should be used for this analyzer? See note (5).

 `NetAnalyzers\Core\Microsoft.NetCore.Analyzers\Runtime`, you might want to check our [instruction for Netcore analyers](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/NetCore_GettingStarted.md)

> Notes:
> (1) I'm assuming this analyzer is also must be applied to C# 9 Records (and structs). This is a valid usage:
> > ```cs
> > public record Person
> > {
> >    public string FirstName { get; init; }
> >    public string LastName { get; init; }
> >
> >    [JsonConstructor]
> >    public Person(string firstName, string lastName)
> >      => (FirstName, LastName) = (firstName, lastName);
> > }
> > ```

If the attribute can be applied for them then the analyzer also should cover them

(2) I'm assuming this analyzer should also be applied to VB.

Yes, the analyzer can be implemented using language-agnostic APIs, so most likely you don't need to do anything extra for VB except adding VB tests

(4) It seems like uninitialized public properties are not a responsibility of this analyzer. Example:
> ```cs
> public class C
> {
>     public int P1 { get; }
>     public int? P2 { get; }
> 
>     [JsonConstructor]
>     public C(int p1)
>     {
>         this.P1 = p1;
>         // P2 is not initialized
>     }
> }
> ```

I think so, i assume we only warn if parameter `p1` haven't had matching property cc @bartonjs

**Comment by psxvoid** [comment]
@buyaa-n Thanks a lot! Then I'll omit fields verification in the analyser if no additional feedback is provided.

 The last question I still have is related to naming styles. Currently, it's possible to set custom prefixes for field/property names in `.editorconfig`, e.g. set `_` (underscore) prefix for private properties. It seems like the analyser should be able to read such options. Without reading and respecting such options a group of users that have set a custom naming style for their projects can get false-positives, e.g. a property `_data` will not much a constructor parameter `data`. For example, here is an article that demonstrates the ability to set a custom prefix, suffix, word-separator and capitalization: [Configure Visual Studio to Name Private Fields with Underscore | ardalis.com](https://ardalis.com/configure-visual-studio-to-name-private-fields-with-underscore/).
 
However, naming style options (code/classes responsible for them) are only available in the `Roslyn` repository and not in `RoslynAnalyzers`. It may be a bit harder to "port" them because they are internal to the `Roslyn` repo and have a lot of internal dependencies. I had spent some time "porting" them (copy-paste, and it seems like a very time-consuming operation) and wondering whether someone from the Roslyn team can recommend a standard and more effective way of doing it (e.g. marking `IPerLanguageOption2` and `NamingStyleOptions` classes `public` in `Roslyn` repo).

**Comment by buyaa-n** [comment]
> @buyaa-n Thanks a lot! Then I'll omit fields verification in the analyser if no additional feedback is provided.

Sorry i was about to update my comment, turns out it is related to fields: `Each parameter in constructor 'Void .ctor(...) on type 'XYZ' must bind to an object property or field on deserialization. Each parameter name must match with a property or field on the object. The match can be case-insensitive.` So if no matching property we can check for fields match.

Another thing we could check for field might be `[JsonInclude]` attribute, but the attribute is not required in case  `JsonSerializerOptions.IncludeFields` set to true, but in most cases, we cannot check if that option is set, so that might be a bit  noisy

**Comment by buyaa-n** [comment]
> The last question I still have is related to naming styles. Currently, it's possible to set custom prefixes for field/property names in .editorconfig, e.g. set _ (underscore) prefix for private properties. It seems like the analyser should be able to read such options. Without reading and respecting such options a group of users that have set a custom naming style for their projects can get false-positives, e.g. a property _data will not much a constructor parameter data. For example, here is an article that demonstrates the ability to set a custom prefix, suffix, word-separator and capitalization: Configure Visual Studio to Name Private Fields with Underscore | ardalis.com.

Is `JsonSerializer` account/respects that?  Can the serialzer set property `_data` with constructor parameter `data` when such styling defined? If it can serialize then yes it is false positive and we need to account styling, but if it cannot then the analyzer should still flag that i think CC @bartonjs @layomia

**Comment by psxvoid** [comment]
> Is `JsonSerializer` account/respects that? Can the serializer set property `_data` with constructor parameter data when such styling defined? 

Yes, I'm sorry; it seems like `JsonConstructor` doesn't account for that and throws an `InvalidOperationException` with the following text:

`Each parameter in constructor 'Void .ctor(xxx)' on type 'yyy' must bind to an object property or field on deserialization.`

It means, as long as the behaviour of the `JsonConstructor` attribute doesn't change, `NamingStyleOptions` can be ignored. Thank you for helping me to understand that.

But this still can be an issue:
> If there is matched field another thing we could flag for field might be `[JsonInclude]` attribute, but the attribute is not required in case `JsonSerializerOptions.IncludeFields` set to true, but in most cases, we cannot check if that option is set, so that might be a bit noisy.

I'm not sure whether there is an infrastructure to share options between .NET packages, but it may be a nice one to have.

**Comment by buyaa-n** [comment]
> It means, as long as the behaviour of the `JsonConstructor` attribute doesn't change, NamingStyleOptions can be ignored. 

Right, thanks for checking that

> I'm not sure whether there is an infrastructure to share options between .NET packages, but it may be a nice one to have.

Sorry, it is not an `infrastructure` option it is `JsonSerializerOptions` can be set/used during serialization 
```cs
   var options = new JsonSerializerOptions
   {
        IncludeFields = true,
   };
   var forecast = JsonSerializer.Deserialize<Foo>(json, options);
```
Which mostly will not be available for static analysis of the declaration of a POCO, that is why i said `most cases, we cannot check if that option is set`

**Comment by psxvoid** [comment]
> Sorry, it is not an infrastructure option it is `JsonSerializerOptions` that can be set/used during serialization

Sorry, for confusing you, I totally get the point. Though `JsonSerializerOptions` can be defined in code, and Roslyn can compile the code on the fly and get those options, most likely it shouldn't be done in the scope of this analyzer. Because `JsonSerializerOptions` can also be defined in different ways, e.g. in AspNetCore in the `Startup.cs`:

```csharp
public void ConfigureServices(IServiceCollection services)
{
	services.AddControllers().AddJsonOptions((JsonOptions o) =>
	{
		o.JsonSerializerOptions.IgnoreNullValues = true;
	});
}
```

By "infrastructure" I meant that it would be nice to have some kind of a unified way of sharing package options (like `JsonSerializationOptions` or `NamingStyleOptions`) between different .NET packages without involving .editorconfig or other external files. But I was just thinking out loud and it's not directly related to this issue, sorry once again.

**Comment by bartonjs** [comment]
I might be wrong here, but I believe that the JsonSerializer/JsonConstructor behavior for fields only considers public fields.  So the analyzer can compare ctor parameter names to field names directly, and ctor parameter names to property names with a camel-to-Pascal conversion.

I don't think there should be any warning about unset properties... they could be just doing calculations.

```C#
public class SomeClass
{
    public string PropOne { get; }

    [JsonConstructor]
    public SomeClass(string propOne)
    {
        PropOne = propOne;
    }

    public int PropTwo => PropOne.Length;
}
```

It's only "hey, there's a ctor parameter here that I'll never figure out how to populate, since it doesn't map to anything that would have been chosen as an automatic name in serialization" that should warn.

```C#

```C#
public class SomeClass
{
    public string PropOne { get; }

    [JsonConstructor]
    public SomeClass(string propOne, int otherParam)
    {
        PropOne = propOne;
    }

    public int PropTwo => PropOne.Length;
}
```

There's not enough data to know where an `otherParam` value should come from, so the type isn't automagically constructable.

**Comment by psxvoid** [comment]
> I might be wrong here, but I believe that the JsonSerializer/JsonConstructor behaviour for fields only considers public fields. So the analyzer can compare ctor parameter names to field names directly, and ctor parameter names to property names with a camel-to-Pascal conversion.

It seems you are correct. Let's consider an example:

```csharp
public class Person
{
    public int age;

    [JsonConstructor]
    public Person(int age)
    {
        this.age = age;
    }
}
```

We will get `InvalidOperationException` if we attempt to deserialize such class. But if we add `IncludeFields` option (or add `JsonInclude` attribute to the `age` field):

```csharp
var options = new JsonSerializerOptions { IncludeFields = true };
```

Then the class will be deserialized correctly without any runtime errors.

But for **private** fields the `InvalidOperationException` will be thrown every time on deserialization despite the presence of the options/attribute. I'm wondering whether we have to take it into account in a code-fix (as well as add another diagnostic message). The code-fix for such a case may recommend making the field public instead of renaming it. But public fields are not recommended by some best-practices, and it may be better to suggest to convert such a field into public property.


> It's only "hey, there's a ctor parameter here that I'll never figure out how to populate, since it doesn't map to anything that would have been chosen as an automatic name in serialization" that should warn.

This situation is most likely the responsibility of `ReviewUnusedParametersAnalyzer`. But because it doesn't responsible for **runtime** behaviour, we may want to handle it in this issue. In such a case, it may be reasonable to add yet another code fix and diagnostic message to this analyzer.

Thank you for revealing those cases.

**Comment by buyaa-n** [comment]
> This situation is most likely the responsibility of ReviewUnusedParametersAnalyzer. But because it doesn't responsible for runtime behaviour, we may want to handle it in this issue. In such a case, it may be reasonable to add yet another code fix and diagnostic message to this analyzer.

@psxvoid I think the developer could have left a parameter without using/setting it on purpose. Plus JsonSerializer does not throw in that case, so i don't think we should warn for unused parameter, especially with a `warning` severity level

**Comment by psxvoid** [comment]
@buyaa-n @genlu Hi!

Sorry, it took a bit longer than "soon" to come back to this issue. May I jump in and continue the implementation?

I saw @genlu as an assignee to the PR and I'm not sure what does it mean.

**Comment by buyaa-n** [comment]
> Sorry, it took a bit longer than "soon" to come back to this issue. May I jump in and continue the implementation?

Sure, please continue the implementation, thank you!

> I saw @genlu as an assignee to the PR and I'm not sure what does it mean.

We are assigned to your PR for review only, we will review your PR when you finish the implemenation

**Comment by SoggyBottomBoy** [comment]
If a property is tagged with the JsonPropertyName which has a value matching the constructor param name, should this not find a match?

```csharp
public class Example
{
    [JsonConstructor]
    public Example(string Name)
    {
        FirstName = Name;
    }

    [JsonPropertyName("Name")]
    public string FirstName { get; }
}

public class ExampleTest
{
    [Test]
    public void JsonRoundTrip()
    {
        var test = new Example("Joe");
        string output = JsonSerializer.Serialize(test);
        var result = JsonSerializer.Deserialize<Example>(output);

        Assert.Equals(test.FirstName, result.FirstName);
    }
}
```

Running the test above gives:
`System.InvalidOperationException: 'Each parameter in the deserialization constructor on type 'Tests.Serialization.Example' must bind to an object property or field on deserialization. Each parameter name must match with a property or field on the object. The match can be case-insensitive.'`

I'm trying to add json attributes to an existing library in which the models (combination of structs and classes), don't always have the property name matching the constructor parameter.

**Comment by eiriktsarpalis** [comment]
@carlossanlop was the `in-pr` label applied correctly? I couldn't a find a PR linked to this issue.

**Comment by buyaa-n** [comment]
The PR is in roslyn-analyzers repo: https://github.com/dotnet/roslyn-analyzers/pull/4877

**Comment by eiriktsarpalis** [comment]
@buyaa-n could you link the PR to this issue? I don't seem to have permissions in the roslyn-analyzers repo.

**Comment by carlossanlop** [comment]
@eiriktsarpalis https://github.com/dotnet/roslyn-analyzers/pull/4877

**Comment by carlossanlop** [comment]
Someone please correct me if I'm wrong: I don't think we can link issues to PRs across repos. That's why I had to add the label manually.

**Comment by eiriktsarpalis** [comment]
Yeah I think it's only possible if you explicitly add a `Fix <link>` statement in the OP of the PR.

**Comment by buyaa-n** [comment]
As there were no progress in the [PR](https://github.com/dotnet/roslyn-analyzers/pull/4877) moving into future milestone, CC @psxvoid.
